### ISSUE TITLE
docker attach not exiting when container exits

### ISSUE THREAD

**Issue by phemmer** [docker/docker#9860]
Very simple test:

    docker attach $(docker run -tid busybox sh -c 'sleep 2; echo exiting')

The command will show the "exiting" message, but never exits. The container is in fact dead, as doing a `docker ps` shows it gone. However pressing any key will then cause the attach to exit.

```
# docker version
Client version: 1.4.1
Client API version: 1.16
Go version (client): go1.3
Git commit (client): 5bc2ff8
OS/Arch (client): linux/amd64
Server version: 1.4.1
Server API version: 1.16
Go version (server): go1.3
Git commit (server): 5bc2ff8

# docker info
Containers: 74
Images: 47
Storage Driver: btrfs
Execution Driver: native-0.2
Kernel Version: 3.18.1-gentoo
Operating System: Gentoo/Linux
CPUs: 4
Total Memory: 7.512 GiB
Name: gadget
ID: KLI3:BPWQ:CK45:NJCL:ULCH:BV6Z:XDFM:YKTZ:EZ7D:UBE7:ZQCK:HOGN
Username: phemmer
Registry: [https://index.docker.io/v1/]
```

**Comment by acbodine** [comment]
Can this be closed as a non-issue?

**Comment by LK4D4** [comment]
@acbodine Is this non-issue?

**Comment by duglin** [comment]
It appears to be waiting on the io.Copy here:
https://github.com/docker/docker/blob/master/api/client/hijack.go#L203

**Comment by acbodine** [comment]
In your example `docker attach $(docker run -tid busybox sh -c 'sleep 2; echo exiting')` you are explicitly keeping your container's `stdin` stream open with the `-i` flag to `docker run`, and excluding the wrapping `docker attach`, your newly created container will sleep for 2 seconds and then print 'exiting' to it's `stdout`.

By looking through the source, when you run a `docker attach` the cli first queries the daemon for the json config of the container you are attempting to attach to. By looking at whether or not the container was created with a setting to keep `stdin` open or not (i.e. `-i` flag to `docker run`) the cli will either setup a client-side pipe to the container's `stdin` or it will not. If it does set up the client-side pipe then there are go routines in the daemon that are stuck waiting for input from the cli to forward on to the container you are attaching to. Hence why the cli is hanging and when you press a key, it will return to a normal state.

Honestly I'm still not sure if this is an issue. I did find that if you add the `--no-stdin=true` flag to your `docker attach` the cli will behave as you expected by explicitly not opening up the client-side pipe to the container's `stdin` regardless of the container config. Otherwise by removing the `-i` flag from your `docker run` - as @LK4D4 has mentioned - the attach should return properly as well.

I still think that maybe it is not an issue because the `-i` flag is intended if you want the container you are creating to accept input on it's `stdin` stream. As such the cli behaves properly by setting up a pipe for you to provide input to the container's `stdin` that you are attaching to. If the container wasn't intended to accept input on it's `stdin` then don't provide the `-i` flag to `docker run`

Thoughts?

**Comment by phemmer** [comment]
> As such the cli behaves properly by setting up a pipe for you to provide input to the container's stdin that you are attaching to.

I'm not saying the attach shouldn't be reading from STDIN. It should be. The container was launched with `-i`, and this is what it's supposed to do.

> there are go routines in the daemon that are stuck waiting for input from the cli to forward on to the container you are attaching to. Hence why the cli is hanging and when you press a key, it will return to a normal state.

But the `docker attach` is also listens for output from the container. While it listens for output, it should also be watching for the container to exit. When it receives that exit event, it should stop waiting for input and exit, returning control to the parent.

> If the container wasn't intended to accept input on it's stdin then don't provide the -i flag to docker run

In my example, `sh -c 'sleep 2; echo exiting'` is an arbitrary process. It could be replaced with anything, maybe something that can read STDIN, but it exited without doing so.

&nbsp;

But just for arguments sake, lets run an interactive command. Try doing the following:

    docker attach $(docker run -tid busybox sh)

Type a few things, run some commands, see that it works, then run `exit`. The process will exit, but you won't be returned to the shell you launched `docker attach` from (until you press another key).

Now what if you're attached to a container and you're waiting for it to do something, but instead it exits without a message. You're still going to be sitting at your `docker attach` waiting for it to do something, not knowing that it's not even running any more.

**Comment by LK4D4** [comment]
Also there is more severe bug about attach not detecting container events: #9797

**Comment by acbodine** [comment]
@phemmer I have seen what you are talking about, sorry for my confusion. Think I found the culprit... working on a pull request now.

**Comment by icecrime** [comment]
Closed by #10079.

**Comment by jfrazelle** [comment]
can confirm this is fixed by #10079
